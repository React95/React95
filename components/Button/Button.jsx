import React from 'react';
import PropTypes from 'prop-types';

import Btn from '../shared-style/Btn';

const Button = ({ children, value, onClick }) => (
  <Btn value={value || children} onClick={onClick}>
    {children || value}
  </Btn>
);

Button.propTypes = {
  value: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  onClick: PropTypes.func,
};

Button.defaultProps = {
  value: 'Ok',
  children: null,
  onClick: () => {},
};

export default Button;
