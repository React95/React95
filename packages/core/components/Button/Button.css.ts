import { style } from '@vanilla-extract/css';
import { calc } from '@vanilla-extract/css-utils';

import { contract } from '../themes/contract.css';

export const button = style({
  backgroundColor: contract.colors.material,
  paddingTop: contract.space[7],
  paddingInline: contract.space[20],
  paddingBottom: contract.space[5],
  border: 'none',
  color: contract.colors.materialText,

  fontSize: 12,

  boxShadow: `inset 1px 1px 0px 1px ${contract.colors.borderLightest},
    inset 0 0 0 1px ${contract.colors.borderDark},
    1px 1px 0 0px ${contract.colors.borderDarkest}`,

  ':disabled': {
    color: contract.colors.materialTextDisabled,
  },
  ':focus': {
    outline: `${contract.space[1]} dotted ${contract.colors.materialText}`,
    outlineOffset: calc.negate(contract.space[5]),

    boxShadow: `inset 1px 1px 0px 1px ${contract.colors.borderLightest},
      inset -0.5px -0.5px 0px 1px ${contract.colors.borderDark},
      1px 1px 0 1px ${contract.colors.borderDarkest}`,
  },
  ':active': {
    paddingTop: contract.space[8],
    paddingInline: contract.space[20],
    paddingBottom: contract.space[4],

    outline: `${contract.space[1]} dotted ${contract.colors.borderDarkest}`,
    outlineOffset: calc.negate(contract.space[5]),

    boxShadow: `inset 0 0 0 1px ${contract.colors.borderDark},
      0 0 0 1px ${contract.colors.borderDarkest}`,
  },
});
