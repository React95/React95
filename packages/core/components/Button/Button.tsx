import React from 'react';
import type {
  ElementType,
  ReactElement,
  ElementRef,
  ForwardedRef,
  ButtonHTMLAttributes,
} from 'react';
import cn from 'classnames';

import { button } from './Button.css';
import { Frame, FrameProps, fixedForwardRef } from '../Frame/Frame';

export type ButtonProps<TAs extends ElementType> =
  ButtonHTMLAttributes<HTMLButtonElement> & FrameProps<TAs>;

export const Button = fixedForwardRef<HTMLButtonElement, ButtonProps<'button'>>(
  (props, ref) => {
    return (
      <Frame
        as="button"
        {...props}
        className={cn(button, props.className)}
        ref={ref}
      />
    );
  },
) as <TAs extends ElementType = 'button'>(
  props: ButtonProps<TAs> & { ref?: ForwardedRef<ElementRef<TAs>> },
) => ReactElement;
