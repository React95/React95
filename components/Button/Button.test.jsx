import React from 'react';
import Button from './Button';
import { render, fireEvent } from 'react-testing-library';

describe('<Button />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { container } = render(<Button />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('value prop', () => {
    it('should display value as Button children', () => {
      const { container } = render(<Button value="Cancel" />);

      expect(container.querySelector('button').textContent).toBe('Cancel');
    });
  });

  describe('onClick prop', () => {
    it('should call onClick function when Button is clicked', () => {
      const onClickMock = jest.fn();
      const { getByText } = render(<Button onClick={onClickMock} value="ok" />);

      fireEvent.click(getByText('ok'));

      expect(onClickMock).toHaveBeenCalled();
    });
  });
});
