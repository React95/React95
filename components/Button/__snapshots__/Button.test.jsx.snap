// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Snapshots should match snapshot 1`] = `
.c0 {
  background-color: #c3c7cb;
  padding: 7px 20px 5px;
  border: none;
  font-size: 12px;
  box-shadow: inset 1px 1px 0px 1px #ffffff,inset 0 0 0 1px #868a8e,1px 1px 0 0px #000;
}

.c0:disabled {
  color: #868a8e;
}

.c0:focus {
  outline: none;
  box-shadow: inset 1px 1px 0px 1px #ffffff,inset -0.5px -0.5px 0px 1px #868a8e,1px 1px 0 1px #000;
  outline: 1px dotted #000;
  outline-offset: -5px;
}

.c0:active {
  padding: 8px 20px 5px;
  outline: 1px dotted #000;
  outline-offset: -5px;
  box-shadow: inset 0 0 0 1px #868a8e,0 0 0 1px #000;
}

<div>
  <button
    class="c0"
    value="Ok"
  >
    Ok
  </button>
</div>
`;
